src/Agent.Worker/TaskDefinitionManager.cs

```diff
@@ -0,0 +1,291 @@
+            ArgUtil.NotNullOrEmpty(task.Version, nameof(task.Version));
+            return Path.Combine(
+                IOUtil.GetTasksPath(HostContext),
+                $"{task.Name}_{task.Id}",
+                task.Version);
+        }
+
+        public async Task DownloadAsync(IExecutionContext executionContext, IEnumerable<TaskInstance> tasks)
+        {
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNull(tasks, nameof(tasks));
+
+            //remove duplicate and disabled tasks
+            IEnumerable<TaskInstance> uniqueTasks =
+                from task in tasks
+                where task.Enabled
+                group task by new
+                    {
+                        task.Id,
+                        task.Version
+                    }
+                into taskGrouping
+                select taskGrouping.First();
+            foreach (TaskInstance task in uniqueTasks)
+            {
+                await DownloadAsync(executionContext, task);
+            }
+        }
+
+        public Definition Load(TaskReference task)
+        {
+            // Validate args.
+            Trace.Entering();
+            ArgUtil.NotNull(task, nameof(task));
+
+            // Initialize the definition wrapper object.
+            var definition = new Definition() { Directory = GetDirectory(task) };
+
+            // Deserialize the JSON.
+            string file = Path.Combine(definition.Directory, Constants.Path.TaskJsonFile);
+            Trace.Info($"Loading task definition '{file}'.");
+            string json = File.ReadAllText(file);
+            definition.Data = JsonConvert.DeserializeObject<DefinitionData>(json);
+
+            // Replace the macros within the handler data sections.
+            foreach (HandlerData handlerData in (definition.Data?.Execution?.All as IEnumerable<HandlerData> ?? new HandlerData[0]))
+            {
+                handlerData?.ReplaceMacros(definition);
+            }
+
+            return definition;
+        }
+    }
+
+    public sealed class Definition
+    {
+        public DefinitionData Data { get; set; }
+        public string Directory { get; set; }
+    }
+
+    public sealed class DefinitionData
+    {
+        public TaskInputDefinition[] Inputs { get; set; }
+        public ExecutionData Execution { get; set; }
+    }
+
+    public sealed class ExecutionData
+    {
+        private readonly List<HandlerData> _all = new List<HandlerData>();
+        private NodeHandlerData _node;
+        private ProcessHandlerData _process;
+
+        [JsonIgnore]
+        public List<HandlerData> All => _all;
+
+        public NodeHandlerData Node
+        {
+            get
+            {
+                return _node;
+            }
+
+            set
+            {
+                _node = value;
+                Add(value);
+            }
+        }
+
+#if !OS_WINDOWS
+        [JsonIgnore]
```

Comment:
This is a pretty slick feature of Json.Net. The execution section for Process handler will only be deserialized on Windows :)